liver_imaging_analysis/models/liver_segmentation.py

```diff
@@ -680,75 +682,85 @@ def train_liver(
+        The frequency at which per_batch_callback will be called.
         Expects a number of epochs. Default is 100.
     save_weight : bool
         whether to save weights or not. Default is True.
     save_path : str
         the path to save weights at if save_weights is True.
-        If not defined, the potential cp path will be loaded 
+        If not defined, the potential cp path will be loaded
         from config.
     test_batch_callback : bool
         whether to call per_batch_callback during testing or not.
         Default is False
     """
-    if cp_path is None:
-        cp_path = config.save["potential_checkpoint"]
-    if epochs is None:
-        epochs = config.training["epochs"]
-    if save_path is None:
-        save_path = config.save["potential_checkpoint"]
+
     set_seed()
     model = LiverSegmentation(modality, inference)
     model.load_data()
+
+    # checkpoints save path
+    cp_dir = os.path.join(
+        config.save["potential_checkpoint"],
+        config.name["experiment_name"],
+        config.name["run_name"],
+        "",
+    )
+    # Check if the directory exists and create it if not
+    if not os.path.exists(cp_dir):
+        os.makedirs(cp_dir)
+    save_path = f"{cp_dir}/potential_checkpoint"
+
     model.data_status()
-    if pretrained:
-        model.load_checkpoint(cp_path)
     model.compile_status()
-    init_loss, init_metric = model.test(
-                                model.test_dataloader, 
-                                callback = test_batch_callback
-                                )
-    print(
-        "Initial test loss:", 
-        init_loss,
-        )
-    print(
-        "\nInitial test metric:", 
-        init_metric.mean().item(),
-        )
-    model.fit(
-        epochs = epochs,
-        evaluate_epochs = evaluate_epochs,
-        batch_callback_epochs = batch_callback_epochs,
-        save_weight = save_weight,
-        save_path = save_path
-    )
-    # Evaluate on latest saved check point
-    model.load_checkpoint(save_path)
-    final_loss, final_metric = model.test(
-                                model.test_dataloader, 
-                                callback = test_batch_callback
-                                )
-    print(
-        "Final test loss:", 
-        final_loss,
+
+    if automate == True:
+        # Create an Optuna study
+        study = optuna.create_study(direction="minimize")
+        study.optimize(
+            lambda trial: model.objective(
+                trial=trial,
+                automate=automate,
+                pretrained=pretrained,
+                cp_path=cp_path,
+                epochs=epochs,
+                evaluate_epochs=evaluate_epochs,
+                batch_callback_epochs=batch_callback_epochs,
+                save_weight=save_weight,
+                save_path=save_path,
+                test_batch_callback=test_batch_callback,
+            ),
+            n_trials=5,
         )
```

Comment:
this should be moved to the engine, it could be used by any model

Reply:
The objective function is already in the engine, or do you mean the study object initialization?

Reply:
yes the study should be initialized in the engine too instead of repeating the same initialization in each model, right?

Reply:
done